### PR TITLE
Icon buttons!

### LINKED ISSUE #216: Button: Allow for alternating direction for navigation buttons
Because navigation is often a case of "go back to whatever", and this makes absolutely no sense:

Screenshot 2020-03-31 at 21 08 57

We probably don't strictly need up and down, but it would be nice to allow for all directions, because why not.

Also this should definitely be using an icon, which it is currently not. Might be able to sneak it in with the icon-button changes happening in #215.

### PR DESCRIPTION
**Pull request by sarahmonster**
Ready for review, finally! It only took me over a month. 😜 

There are a number of changes here, so good to take it for a spin and really test it out to get a feel for how it works in practise.

Screenshot 2020-05-08 at 15 54 28

Screenshot 2020-05-08 at 15 54 45

Here are the major changes:
- Added an "iconDirection" prop to allow for overwriting the default right-facing arrow. I have kinda mixed feelings about this name, since it could imply that it works for any icon and flips the icon, whereas it's only intended to be used to change the navigation button icon, for cases where your navigational direction is actually different. (So "go back" would warrant a left-facing arrow, for instance.) It's really determining the direction of the _navigation_ action, not the icon. What about `navigationDirection` instead? (And have back/forward/up/down as options?)
- Added the ability to set both leading and trailing icons to any button.
- Refactored semantic and navigation buttons to use proper icons.
- Added the ability to set stackedIcon or onlyIcon, for stacked icon+text buttons, or icon-only buttons.
- Improved spacing and alignment of icons inside buttons.

Before:
Screenshot 2020-05-08 at 15 55 19

After:
Screenshot 2020-05-08 at 15 54 38

Things that we may want to do prior to pushing:
- Update tests to add more tests.
- Check to see that passing a null value to semantic or navigation buttons will nix the icon. (I tried this by passing `<Button navigation trailingIcon={null} ` but I might have the syntax wrong?)
- Update README to better reflect changes. The button README is one of the first ones we wrote and thus really needs a total overhaul, so I'm kinda tempted to do it in one fell swoop after some other button improvements I had in mind. We could definitely sketch out the pertinent updates in this PR, though, and then refine later.
- I'm realising I've made the icons a good deal bigger... do we want to go with this, or should they be smaller when used as a leading or trailing icon? (stacked and onlyicon buttons definitely feel like the right size.)
- Re-animate the navigation button icons. I'll do this once we've determined the API!

Fix #216.

### CODE REVIEW

File: src/components/ui/Button/README.md [hunk 1]
```diff
@@ -42,9 +42,11 @@ Use a primary button to draw attention to the most important action on the page.
 Minimal buttons are good when you don't want to draw too much attention to a button. Use them for secondary actions.
 
 ```jsx
-<Button primary>Sign up now</Button>
-<Button>Sign up sometime</Button>
-<Button minimal>Sign up eventually</Button>
+<div style={{display: 'flex', flexWrap: 'wrap'}}>
```
Comment (tofumatt) [inline src/components/ui/Button/README.md]: I wonder if we should have a component for this in the styleguide? Or can we use emotion's `css` prop for it instead? I'll give it a go; hardly a big deal!

File: src/components/ui/Button/README.md [hunk 2]
```diff
@@ -64,17 +66,39 @@ A link loses focus and means the user needs to reorient themselves, but a button
...
+
+  <Button navigation iconDirection="left" href="https://twitter.com/KittenBreak">
+    Go left
+  </Button>
+
+  <Button navigation iconDirection="up" href="https://twitter.com/KittenBreak">
+    Go up
+  </Button>
+
+    <Button navigation iconDirection="down" href="https://twitter.com/KittenBreak">
```
Comment (tofumatt) [inline src/components/ui/Button/README.md]: Odd alignment here, but I'll fix it.

File: src/components/ui/Button/README.md [hunk 3]
```diff
@@ -128,19 +158,109 @@ Use a danger button any time you have a destructive action, like deleting someth
...
+    Minimal disabled
+  </Button>
+</div>
 ```
 
 What's my button up to? State will tell you!
+
+### Icon
+
+Use icons in buttons to reinforce meaning.
```
Comment (tofumatt) [inline src/components/ui/Button/README.md]: I think this section is a bit out-of-date and not very clear. Could you go back to it and improve this documentation? I think it's a bit disconnected from the code examples.

Screenshot 2020-05-08 16 39 38

**Comment by tofumatt** [comment]
Oh, regarding `navigationDirection`… I want to make it clear that the direction is for the icon, so let's call it `navigationIcon` and just allow for `"left" | "right" | "up" | "down"` as props. I think that's fine!

Also: I like the bigger icon sizes! :shipit:

File: src/components/ui/Button/README.md (hunk 1 above)
Comment (sarahmonster) [reply src/components/ui/Button/README.md]: This works! There was an error in your CSS that caused wrapping issues, but I've fixed it. Thanks!

File: src/components/ui/Button/README.md (hunk 3 above)
Comment (sarahmonster) [reply src/components/ui/Button/README.md]: Agreed! I'll run through the docs and see about improving them prior to merging—they could use a lot of ❤️

**Comment by sarahmonster** [comment]
Got distracted and decided to make our semantic buttons a bit more semantic, using shape to underscore their meaning:

Screenshot 2020-05-08 at 17 49 42

File: src/components/ui/Button/README.md (hunk 3 above)
Comment (sarahmonster) [reply src/components/ui/Button/README.md]: That was actually definitely some notes to self there. Docs have been reworked, fleshed out, and rearranged now—let me know how it works!

File: src/components/ui/Button/README.md [hunk 4]
```diff
@@ -7,52 +7,130 @@ const onClick = (event) => {
...
+
+There are several different permutations of icon buttons available:
 
-- Only use a single **primary** button per screen.
-- Use prominence to **highlight the most important action** a user can take.
-- When using two or more buttons together, try to use different prominence levels to guide users to the most likely actions.
+- Leading Icon: shows an icon before the text.
+- Trailing Icon: shows an icon after the text.
+- Stacked Icon: shows an icon with smaller text underneath. Useful for application toolbars.
+- Only Icon: shows an icon only, in a rounded container. Useful in areas of limited space where the icon's meaning is very clear.
```
[the remark this answers is not in the document]
Comment (sarahmonster) [reply src/components/ui/Button/README.md]: Agreed! I'm actually just mirroring the props here—maybe we should change those to match?

File: src/components/ui/Button/README.md [hunk 5]
```diff
@@ -64,83 +142,125 @@ A link loses focus and means the user needs to reorient themselves, but a button
...
-  Woke up on time
-</Button>
+import { Flex } from 'nautilus-styleguide';
+
+<Flex>
+  <Button primary success>
+    Nobel Peace Prize
+  </Button>
+  <Button success>Got a new job</Button>
+  <Button minimal success>
```
[the remark this answers is not in the document]
Comment (sarahmonster) [reply src/components/ui/Button/README.md:203]: I think you did this already—thanks!

File: src/components/ui/Button/index.js [hunk 6]
```diff
@@ -45,6 +55,46 @@ const Button = (props) => {
 
   let Component = 'button';
   let linkPropsToUse;
+  let trailingIconName;
+  let leadingIconName;
+
+  // Set directional icons for navigation buttons.
```
[the remark this answers is not in the document]
Comment (sarahmonster) [reply src/components/ui/Button/index.js:58]: Probably! Buttons are COMPLICATED.

File: src/components/ui/Button/README.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (sarahmonster) [reply src/components/ui/Button/README.md]: Actually apparently the prop is iconOnly and I'm a real dummy.

I nixed your hyphen though.

**Comment by sarahmonster** [comment]
Okay! They're still not perfect, but I think a lot better. Good catch with the focus styles, there were a few things going on there that would prove to be problematic that I've patched.

We'll probably want to do a little bit of housekeeping with the interactive tokens (how did I not know these were here!?) in the future to make it more themeable and flexible, but I think that'll be cause for another, larger PR—and easier to tell what we really need once we've had a chance to work with the icon buttons a bit more!

**Pull request merged**